client/my-sites/plans-features-main/hooks/use-plan-types-with-intent.ts

```diff
@@ -0,0 +1,105 @@
+import {
+	TYPE_BLOGGER,
+	TYPE_BUSINESS,
+	TYPE_ECOMMERCE,
+	TYPE_ENTERPRISE_GRID_WPCOM,
+	TYPE_FREE,
+	TYPE_PERSONAL,
+	TYPE_PREMIUM,
+	getPlan,
+	isBloggerPlan,
+} from '@automattic/calypso-products';
+
+export type Intent =
+	| 'blog-onboarding'
+	| 'newsletter'
+	| 'link-in-bio'
+	| 'new-hosted-site'
+	| 'new-hosted-site-hosting-flow'
+	| 'plugins'
+	| 'jetpack-app' // newly added
+	| 'import'
+	| 'default';
+
+interface Props {
+	intent: Intent;
+	selectedPlan: string;
+	sitePlanSlug: string;
+	hideEnterprisePlan: boolean;
+}
+
+interface PlanTypesWithIntent {
+	intent: Intent;
+	planTypes: string[];
+}
+
+const usePlanTypesWithIntent = ( props: Props ): PlanTypesWithIntent | null => {
+	const { selectedPlan, sitePlanSlug, hideEnterprisePlan, intent } = props;
+	const isBloggerAvailable = isBloggerPlan( selectedPlan ) || isBloggerPlan( sitePlanSlug );
+
+	// TODO:
+	// this should fall into the processing function for the visible plans
+	// however, the Enterprise plan isn't a real plan and lack of some required support
+	// from the utility functions right now.
+	const isEnterpriseAvailable = ! hideEnterprisePlan;
+
+	let currentSitePlanType = null;
+	if ( sitePlanSlug ) {
+		currentSitePlanType = getPlan( sitePlanSlug )?.type;
+	}
+
+	const defaultPlanTypes = [
+		TYPE_FREE,
+		...( isBloggerAvailable ? [ TYPE_BLOGGER ] : [] ),
+		TYPE_PERSONAL,
+		TYPE_PREMIUM,
+		TYPE_BUSINESS,
+		TYPE_ECOMMERCE,
+		...( isEnterpriseAvailable ? [ TYPE_ENTERPRISE_GRID_WPCOM ] : [] ),
+	];
+
+	switch ( intent ) {
+		case 'blog-onboarding':
+			return {
+				intent,
+				planTypes: [ TYPE_FREE, TYPE_PERSONAL, TYPE_PREMIUM, TYPE_BUSINESS ],
+			};
+		case 'newsletter':
+		case 'link-in-bio':
+			return {
+				intent,
+				planTypes: [ TYPE_FREE, TYPE_PERSONAL, TYPE_PREMIUM ],
+			};
+		case 'new-hosted-site':
+			return {
+				intent,
+				planTypes: [ TYPE_FREE, TYPE_PERSONAL, TYPE_PREMIUM, TYPE_BUSINESS, TYPE_ECOMMERCE ],
+			};
+		case 'new-hosted-site-hosting-flow':
+			return {
+				intent,
+				planTypes: [ TYPE_BUSINESS, TYPE_ECOMMERCE ],
+			};
+		case 'import':
+			return {
+				intent,
+				planTypes: [ TYPE_FREE, TYPE_PERSONAL, TYPE_PREMIUM, TYPE_BUSINESS ],
+			};
+		case 'plugins': // is this used?
+			return {
+				intent,
+				planTypes: [
+					...( currentSitePlanType ? [ currentSitePlanType ] : [] ),
+					TYPE_BUSINESS, // this can match the currentSitePlanType. needs investigation.
```

Comment:
I was wondering if you have investigated this already?

Reply:
Nope. Needs to be tested in other PR, if this is even used. Do we render plans in `/plugins`?

Reply:
No, `/plugins` is defaulted to the Business plan. From my test, currently it leads to `/start/business`, so we don't need to worry about it.

Reply:
Sound. will remove the comments.

@southp do we need to revisit anything at all related to plans rendered under /plugins routes? I wasn't able to follow the routes and get the plans view, but haven't tested enough